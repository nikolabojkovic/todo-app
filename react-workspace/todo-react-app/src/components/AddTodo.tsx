import { Button, Col, Container, Form, Row, Stack } from 'react-bootstrap';
import { useState } from 'react';
import { useTodoList, useTodoListDispatch } from '../context/TodosContext';
import { ITodo } from '../models/Todo';

export function AddTodo() {
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');

  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();
  const id = todoList.originalList.length >= 1 
    ? todoList.originalList
        .sort((a: ITodo, b: ITodo) => a.id > b.id ? 1 : -1)[todoList.originalList.length - 1].id 
    : 0;

  return (
    <Form className="todo-background p-1">
      <Container fluid>
        <Row xs={1} sm={3}>
          <Col sm={5} className="p-2">
            <Form.Control 
              type="text" 
              placeholder="Enter title" 
              size="sm" 
              value={title}
              onChange={(e) => setTitle(e.target.value)} 
            />
          </Col>
          <Col sm={5} className="p-2">
            <Form.Control 
              type="text" 
              placeholder="Enter description" 
              size="sm" 
              value={description}
              onChange={(e) => setDescription(e.target.value)} 
            />
          </Col>
          <Col sm={2} className="p-2">
            <Button 
              variant="warning"
              style={{ 
                backgroundColor: '#FE9801',
                color: 'white', 
                minWidth: '90px',
                borderRadius: '20px',
                width: '100%'
              }} 
              size="sm"
              disabled={!title || title.trim() === '' || !description || description.trim() === ''}
              onClick={() => {
                dispatch({
                  type: 'added',
                  id: id + 1, 
                  title, 
                  description
                });
                setTitle('');
                setDescription('');
              }}
            >
              Add
            </Button>
          </Col>
        </Row>
      </Container>
    </Form>
  )
}