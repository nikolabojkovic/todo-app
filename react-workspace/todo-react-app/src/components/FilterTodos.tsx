import { Form, Stack } from "react-bootstrap";
import { useTodoList, useTodoListDispatch } from "../context/TodosContext";

export function FilterTodos() {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  function onFilter(completed: any, uncompleted: any) {
    dispatch({
      type: 'filtered',
      activePage: 1,
      filter: { 
        completed: completed,
        uncompleted: uncompleted
      }
    });
  }

  return (
    <Form className="App__filter d-flex justify-content-start align-items-center">
      <Stack direction="horizontal" gap={3}>       
        <Form.Check 
          type="switch"
          id={`completed`}
          label={`Completed`}
          checked={todoList.filter.completed}
          onChange={(e) => {
            onFilter(e.target.checked, todoList.filter.uncompleted);
          }}
        />   
        <Form.Check 
          type="switch"
          id={`uncompleted`}
          label={`Uncompleted`}
          checked={todoList.filter.uncompleted}
          onChange={(e) => {
            onFilter(todoList.filter.completed, e.target.checked);
          }}
        />         
      </Stack>
    </Form>
  )
}