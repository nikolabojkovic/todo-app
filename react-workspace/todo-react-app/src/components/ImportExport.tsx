import { faFileExport, faFileImport } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useRef, useState } from "react";
import { Button, Col, Container, Form, InputGroup, Row } from "react-bootstrap";
import { useTodoList, useTodoListDispatch } from "../context/TodosContext";
import { Todo } from "../models/Todo";

const buttonStyle : any = { 
  backgroundColor: '#F5F6F7', 
  borderRadius: '20px', 
  minWidth: '90px'
}

const fileReader = new FileReader();

export function ImportExport() {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();
  
  const [file, setFile] = useState(null);
  const fileRef = useRef(null as any);

  function handleImport(e: any) {
    e.preventDefault();

    if (!file) 
      return;

    fileReader.readAsText(file);
  };

  function handleExport() {
    const link = document.createElement('a');
    const jsonContent = `data:text/json;chatset=utf-8,${encodeURIComponent(
      JSON.stringify(todoList.originalList)
    )}`;
    link.href = jsonContent;
    link.download = `todo-list-${new Date().toLocaleDateString()}-${new Date().toLocaleTimeString()}.json`;
    link.click();
  }

  fileReader.onload = function (e: any) {
    const text = e.target.result;
    const list = JSON.parse(text) as Todo[];

    if (!(list instanceof Array)) {
      alert("Invalid JSON file content. Todo list should be an array.");
      return;
    }

    const importedTodoList = list.map((item: any) => new Todo(item.id, 
                                                              item.title, 
                                                              item.description, 
                                                              item.completed, 
                                                              item.createdAt));

    if ((importedTodoList.length > 0 
        && (!(importedTodoList[0] instanceof Todo) 
         || !(Todo.validateFields(importedTodoList[0]))
            )
        )) {
      alert("Invalid JSON file content. Objects in array are not valid Todo objects.");
      return;
    }
          
    setFile(null);
    dispatch({
      type: 'imported',
      originalList: importedTodoList,
      activePage: 1
    });
    fileRef.current.value = '';
  };

  return (
    <Form className="todo-background p-1">
      <Container fluid>
        <Row xs={1} sm={3}>
          <Col sm={8} className="p-2">
            <InputGroup size="sm" className="">
              <Form.Control
                aria-label="Choose File"
                aria-describedby="inputGroup-sizing-sm"
                type={"file"} 
                accept={".json"} 
                ref={fileRef}
                onChange={(e: any) => {
                  setFile(e.target.files[0]);
                }}
                readOnly 
              />
            </InputGroup>
          </Col>
          <Col sm={2} className="p-2">
            <Button 
              className="w-100"
              variant="outline-success" 
              size="sm"
              style={ buttonStyle }
              onClick={handleImport}
              disabled={!file}
            >
              Import {' '}
              <FontAwesomeIcon icon={faFileImport} />
            </Button>
          </Col>
          <Col sm={2} className="p-2">
            <Button 
              className="w-100"
              variant="outline-success" 
              size="sm"
              style={ buttonStyle }
              disabled={!todoList.originalList || todoList.originalList.length === 0}
              onClick={handleExport}
            >
              Export {' '}
              <FontAwesomeIcon icon={faFileExport} />
            </Button>
          </Col>
        </Row>
      </Container>
    </Form>
  )
}