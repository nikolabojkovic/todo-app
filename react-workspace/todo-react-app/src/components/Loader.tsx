export function Loader() {
  return (<div className='loader-container d-flex justify-content-center align-items-center fade-in'>
            <div className='loader'></div>
          </div>);
}