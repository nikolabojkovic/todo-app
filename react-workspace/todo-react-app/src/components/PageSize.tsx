import { Stack, Col } from 'react-bootstrap';
import Dropdown from 'react-bootstrap/Dropdown';
import { useTodoList, useTodoListDispatch } from '../context/TodoListContext';
import { IAction } from '../models/Action';
import { useState } from 'react';

type Props = {
  inputSelectRef: any,
  pageCount: number
};

export function PageSize({ inputSelectRef, pageCount }: Props) {

  const [itemsPerPage, setItemsPerPage] = useState(5);

  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  function handlePageSizeChange(pageSize: number) {
    dispatch({
      type: 'paging-updated',
      payload: {
        activePage: 1,
        itemsPerPage: pageSize
      }
    } as IAction);
    setItemsPerPage(pageSize);
  }
  
  return (
  <>
    <Col sm={3} className="p-3 pt-2 pb-2 d-flex justify-content-sm-start justify-content-center align-items-center">
      Page {todoList.paging.activePage} of {pageCount}
    </Col>
    <Col sm={4} className="p-3 pt-2 pb-2">
      <Stack direction="horizontal" gap={2} className="justify-content-center">
        <div className='ms-sm-auto'>
          Page size:{' '}
        </div>
        <div className="">
          <Dropdown>
            <Dropdown.Toggle
              className="action-button" 
              ref={inputSelectRef} 
              id="dropdown-basic"
            >
              {' '}{itemsPerPage}{' '}
            </Dropdown.Toggle>

            <Dropdown.Menu>
              <Dropdown.Item href="#/action-1" onClick={() => handlePageSizeChange(5)}>5</Dropdown.Item>
              <Dropdown.Item href="#/action-2" onClick={() => handlePageSizeChange(10)}>10</Dropdown.Item>
              <Dropdown.Item href="#/action-3" onClick={() => handlePageSizeChange(50)}>50</Dropdown.Item>
              <Dropdown.Item href="#/action-3" onClick={() => handlePageSizeChange(100)}>100</Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </div>
      </Stack>
    </Col>
  </>)
}