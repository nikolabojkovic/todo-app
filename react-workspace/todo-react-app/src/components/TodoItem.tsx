import { Button, Stack } from 'react-bootstrap';
import { useTodoListDispatch } from '../context/TodoListContext';
import { IAction } from '../models/Action';
import { ITodo } from '../models/Todo';

type Props = {
  todo: ITodo;
};

export function TodoItem({ todo }: Props){

  const dispatch = useTodoListDispatch();

  return (
      <div className="App__todo-list__item">
        <Stack direction="horizontal" gap={3}>
          <div>
            <div className={ todo.completed 
                ? "App__todo-list__item-title--completed" 
                : "App__todo-list__item-title" 
            }>
              { todo.title }
            </div>
            <div className={
              todo.completed 
                ? "App__todo-list__item-description--compleated" 
                : "App__todo-list__item-description"
              }>
                <span>{todo.description}</span>
                <span>
                  {' ' + new Date(todo.createdAt).toDateString() + ' '}
                </span>
            </div>
          </div>     
          { !todo.completed &&
            <Button 
              className="ms-auto confirm-button"
              variant="outline-success" 
              size="sm"
              onClick={() => {
                dispatch({
                  type: 'changed',
                  payload: {
                    todo: {...todo, completed: true}
                  }
                } as IAction);
              }}
            >
              Complete
            </Button>
          }
          <Button 
            className= { todo.completed ? "ms-auto confirm-button" : "confirm-button" } 
            variant="outline-danger" 
            size="sm"
            onClick={() => {
              dispatch({
                type: 'deleted',
                payload: {
                  id: todo.id
                }                
              } as IAction);
            }}
          >
            Delete
          </Button>
        </Stack>
      </div>
  )
}