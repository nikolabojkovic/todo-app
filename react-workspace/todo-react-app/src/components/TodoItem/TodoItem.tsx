import { useState } from 'react';
import { Stack } from 'react-bootstrap';
import { faTrash, faCheckDouble } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useTodoListDispatch } from '../../context/TodoListContext';
import { IAction, TodoActions } from '../../models/Action';
import { ITodo } from '../../models/Todo';
import { ConfirmModal } from '../ConfirmModal/ConfirmModal';

type Props = {
  todo: ITodo;
};

export function TodoItem({ todo }: Props){
  const [showModal, setShowModal] = useState(false); 
  const [confirm, onConfirm] = useState<null | (() => void)>(null);   
  const [confirmDialogText, setConfirmDialogText] = useState(''); 
  const dispatch = useTodoListDispatch();

  function handleComplete() {
    dispatch({
      type: TodoActions.changed,
      payload: {
        todo: {...todo, completed: true}
      }
    } as IAction);
  }

  function handleDelete() {
    dispatch({
      type: TodoActions.deleted,
      payload: {
        id: todo.id
      }
    } as IAction);
  }

  return (
      <div className="App__todo-list__item">
        <Stack direction="horizontal" gap={3}>
          <div>
            <div className={ todo.completed 
                ? "App__todo-list__item-title--completed" 
                : "App__todo-list__item-title" 
            }>
              { todo.title }
            </div>
            <div className={
              todo.completed 
                ? "App__todo-list__item-description--compleated" 
                : "App__todo-list__item-description"
              }>
                <span>{todo.description}</span>
                { todo.createdAt && 
                  <span>
                    {' (' + new Date(todo.createdAt).toDateString() + ')'}
                  </span> 
                }
            </div>
          </div>
          <div className="ms-auto">
            <FontAwesomeIcon 
              data-testid='complete-button'
              icon={faCheckDouble} 
              className={todo.completed ? "action-icon--disabled" : "action-icon"}
              onClick={() => {
                if (todo.completed)
                  return;

                onConfirm(() => handleComplete);
                setConfirmDialogText('Are you sure you want to complete this item?');
                setShowModal(true);
              }}
            />
          </div>
          <div>
            <FontAwesomeIcon 
              data-testid="delete-button"
              icon={faTrash} 
              className="action-icon"
              onClick={() => {
                onConfirm(() => handleDelete);
                setConfirmDialogText('Are you sure you want to delete this item?');
                setShowModal(true);
              }}
            />
          </div>          
        </Stack>
        <ConfirmModal
          content={confirmDialogText} 
          show={showModal}
          onConfirm={() => { 
            ((confirm!)());
            setShowModal(false);
          }}
          onCancel={() => setShowModal(false)}
        />
      </div>
  );
}