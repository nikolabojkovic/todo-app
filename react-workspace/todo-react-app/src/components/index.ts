export * from './TodoSettings';
export * from './Sort';
export * from './TodoList/TodoList';
export * from './Tabs/Tabs';
export * from './Paging/Paging';
export * from './AddTodo/AddTodo';
export * from './ConfirmModal/ConfirmModal';
export * from './Filter/FilterTodos';
export * from './ImportExport/ImportExport';
export * from './Pagination/PageSize';
export * from './Pagination/Pagination';
export * from './Search/Search';
export * from './TodoItem/TodoItem';
export * from './Loader/Loader';
export * from './TodoActions/TodoActions';
export * from './LoadMore/LoadMore';