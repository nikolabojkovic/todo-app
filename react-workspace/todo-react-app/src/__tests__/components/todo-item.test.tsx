import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import { TodoItem } from '../../components/TodoItem';

const todo = {
  id: 1,
  title: "Task 1",
  description: "Description 1",
  completed: false,
  createdAt: new Date(2023, 4, 5)
};

const todoCompleted = {
  id: 1,
  title: "Task 1",
  description: "Description 1",
  completed: true,
  createdAt: new Date(2023, 4, 5)
};

describe('todo item uncompleted', () => {
  it('match snapshot', () => {
    const tree = renderer.create(
      <TodoItem todo={todo}/>
    ).toJSON();
    render(<TodoItem todo={todo}/>);
    expect(tree).toMatchSnapshot();
  });

  it('has 2 buttons', () => {
    render(<TodoItem todo={todo}/>);
    const buttons = screen.getAllByRole('button', { hidden: true });
    expect(buttons.length).toBe(2);
    expect(buttons[0].textContent).toBe('Complete');
    expect(buttons[1].textContent).toBe('Delete');
  });
})

describe('todo item completed', () => {
  it('has completed class', () => {    
    render(<TodoItem todo={todoCompleted}/>);
    const titleElement = screen.getByText(todoCompleted.title);
    expect(titleElement).toHaveClass('App__todo-list__item-title--completed');
  });

  it('has only delete button', () => {
    render(<TodoItem todo={todoCompleted}/>);
    const button = screen.getByRole('button', { hidden: true });
    expect(button.textContent).toBe('Delete');
  });
});

