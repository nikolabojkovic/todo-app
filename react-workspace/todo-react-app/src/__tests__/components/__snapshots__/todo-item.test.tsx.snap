// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo item uncompleted match snapshot 1`] = `
<div
  className="App__todo-list__item"
>
  <div
    className="hstack gap-3"
  >
    <div>
      <div
        className="App__todo-list__item-title"
      >
        Task 1
      </div>
      <div
        className="App__todo-list__item-description"
      >
        <span>
          Description 1
        </span>
        <span>
           Fri May 05 2023 
        </span>
      </div>
    </div>
    <button
      className="ms-auto confirm-button btn btn-outline-success btn-sm"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Complete
    </button>
    <button
      className="confirm-button btn btn-outline-danger btn-sm"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Delete
    </button>
  </div>
</div>
`;
